internal/controller/preservation/handler.go

```diff
@@ -0,0 +1,358 @@
+			return err
+		}
+		log.Info("Successfully deleted", zap.String("resource", resourceType),
+			zap.String("namespace", p.Namespace), zap.String("name", name))
+		return nil
+	}
+
+	var errs []error
+
+	// Delete ConfigMaps
+	for _, name := range data.ConfigMaps {
+		if err := delete(name, configMapType); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	// Delete Secrets
+	for _, name := range data.Secrets {
+		if err := delete(name, secretType); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return errors.NewAggregate(errs)
+}
+
+// buildBackupLabelSelector constructs a label selector for backup operations based on the PreservationMode.
+// Returns nil for PreservationModeNone as no preservation is required.
+func (p *PreservationHandler) buildBackupLabelSelector() (labels.Selector, error) {
+	var (
+		requirement *labels.Requirement
+		err         error
+	)
+
+	switch p.PreservationMode {
+	case v1alpha1.PreservationModeNone:
+		// No preservation required.
+		return nil, nil
+	case v1alpha1.PreservationModeAll:
+		// Select all resources with the PreservationLabelKey.
+		requirement, err = labels.NewRequirement(
+			v1alpha1.PreservationLabelKey,
+			selection.Exists,
+			nil,
+		)
+	case v1alpha1.PreservationModeClusterIdentity:
+		// Select resources with PreservationLabelKey set to ClusterIdentityLabelValue.
+		requirement, err = labels.NewRequirement(
+			v1alpha1.PreservationLabelKey,
+			selection.Equals,
+			[]string{v1alpha1.ClusterIdentityLabelValue},
+		)
+	default:
+		// Handle unknown PreservationMode.
+		return nil, fmt.Errorf("unknown PreservationMode: %v", p.PreservationMode)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create backup label selector: %w", err)
+	}
+
+	// Create and return the label selector with the constructed requirement.
+	return labels.NewSelector().Add(*requirement), nil
+}
+
+// buildRestoreLabelSelector constructs a label selector for restore operations based on the PreservationMode.
+// Returns nil for PreservationModeNone as no preservation is required.
+func (p *PreservationHandler) buildRestoreLabelSelector() (labels.Selector, error) {
+	var (
+		requirement *labels.Requirement
+		err         error
+	)
+
+	switch p.PreservationMode {
+	case v1alpha1.PreservationModeNone:
+		// No restoration required.
+		return nil, nil
+	case v1alpha1.PreservationModeAll, v1alpha1.PreservationModeClusterIdentity:
+		// Select all resources with the backup preservation label.
+		requirement, err = labels.NewRequirement(
+			preservedLabel,
+			selection.Exists,
+			nil,
+		)
+	default:
+		// Handle unknown PreservationMode.
+		return nil, fmt.Errorf("unknown PreservationMode: %v", p.PreservationMode)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create restore label selector: %w", err)
+	}
+
+	// Create and return the label selector with the constructed requirement.
+	return labels.NewSelector().Add(*requirement), nil
+}
```

Comment:
Can you deduplicate these? They look entirely identical except for the actual label (which could be function arg).